File: ios_app/iosApp/Multiplatform/MultiplatformDemo.swift
//
//  MultiplatformDemo.swift
//  iosApp
//
//  Created by Liber Da Silva on 07/08/2020.
//

import Flutter

open class MultiplatformDemo {
    public static var instance = MultiplatformDemo()
    
    var aliveEngines = [String: MultiplatformDemoEngine]()
    
    public var mainEngine: MultiplatformDemoEngine? {
        get {
            getEngineInstance(multiplatformDemoEngine)
        }
    }
    
    public func initialize() throws {
        _ = try startNewEngineInstance(multiplatformDemoEngine)
    }

    public func startNewEngineInstance(_ engineId: String) throws -> MultiplatformDemoEngine {
        return try MultiplatformDemoEngine(engineId)
    }
    
    public func engineInstanceExists(_ engineId: String) -> Bool {
        return aliveEngines[engineId] != nil
    }

    public func getEngineInstance(_ engineId: String) -> MultiplatformDemoEngine? {
        return aliveEngines[engineId]
    }

    public func getOrStartNewEngineInstance(_ engineId: String) throws -> MultiplatformDemoEngine {
        return try getEngineInstance(engineId) ?? startNewEngineInstance(engineId)
    }

    public func destroy() {
        aliveEngines.forEach { aliveEngine in
            aliveEngine.value.stop()
        }
    }

    // MARK: Constants
    let multiplatformDemoEngine = "multiplatformDemoEngine"        
}

File: android_app/app/src/main/java/com/rappipay_demo/android_app/multiplatform/MultiplatformDemo.kt
package com.rappipay_demo.android_app.multiplatform

import android.content.Context
import java.util.concurrent.ConcurrentLinkedQueue

object MultiplatformDemo {
    val aliveEngines = ConcurrentLinkedQueue<String>()
    val mainEngine : MultiplatformDemoEngine?
        get() = getEngineInstance(MULTIPLATFORM_DEMO_ENGINE_ID)

    fun initialize(
        context: Context
    ) {
        if (!engineInstanceExists(MULTIPLATFORM_DEMO_ENGINE_ID)) {
            startNewEngineInstance(context, MULTIPLATFORM_DEMO_ENGINE_ID)
        }
    }

    fun startNewEngineInstance(context: Context, engineId: String) = MultiplatformDemoEngine(context, engineId)

    fun getOrStartNewEngineInstance(context: Context, engineId: String) = getEngineInstance(engineId) ?: startNewEngineInstance(context, engineId)

    fun getEngineInstance(engineId: String) = MultiplatformDemoEngine.getInstance(engineId)

    fun engineInstanceExists(engineId: String) = MultiplatformDemoEngine.instanceExists(engineId)

    fun destroy() {
        aliveEngines.forEach {
            MultiplatformDemoEngine.getInstance(it)?.stop()
        }
    }

    const val MULTIPLATFORM_DEMO_ENGINE_ID = "MULTIPLATFORM_DEMO_ENGINE_ID"
}